proposed/2022/package-status-indicators.md

```diff
@@ -0,0 +1,120 @@
+
+One of the top themes from both the NuGet (Visual Studio) and NuGet.org surveys include "warnings for known package security vulnerabilities and suggested fixes" and "deprecation and vulnerability indicators" averaging 18 points being spent on the feature areas out of 100 total points (7,403pts over 418 responses and 4,405pts 247 responses respectfully). These are two of the top areas that developers would love improvements to. We have confidence that security is on top of developer's minds and they would like to see these experiences in more locations.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+When a developer is working within Visual Studio, they may regularly check their "Dependencies" node within the solution explorer. When expanded, a "Packages" node is available to show the hierarchy of top-level and transitive packages included in a project.
+
+NuGet uses three main words to describe these status of a package, the TL;DR are:
+
+- **Deprecated** - Please don't use this for a disclosed reason and please use a suggested alternative(if any).
+- **Vulnerable** - Please don't use this due to a known vulnerability and update/replace/remove/do nothing accordingly.
+- **Outdated** - Please consider using a newer version.
+
+Building upon Tim's initial mockup, there should be a UI affordance that demonstrates the truth of each package's status in the solution explorer representing these three words.
+
+![](../../meta/resources/PackageStatusIndicators/package-status-indicators.png)
+
+For the sake of not annoying developers, we should only bubble up a warning affordance to the top-level node (Packages and Dependencies) if there is a known top-level vulnerability or deprecated package in their package graph.
+
+Outdated packages would still show the indicators if the packages node is expanded however.
+
+Currently, NuGet only uses two icons to display status indicators:
+
+- A warning triangle (vulnerability and deprecation).
+![](../../meta/resources/PackageStatusIndicators/StatusWarning.png)
+- An up arrow (outdated).
+![](../../meta/resources/PackageStatusIndicators/StatusUpdateAvailable.png)
+
+Hovering over the package and icon will display the respective information for the status.
+
+- Deprecated - "This package has been deprecated as it is `<deprecation category>`."
+- Vulnerable - "This package has at least one vulnerability with `<severity category>` severity."
+- Outdated - "This package has version `<latest version>` available.
+
+When right clicking a package, a developer should be able to get to the "Manager NuGet Packages..." context menu action like they can with the "Packages" and "Dependencies" nodes.
```

Comment:
It would be great if we could have more targeted gestures in these menus for packages. "Manage" is vaguer than "See updates" or "See vulnerabilities".

Reply:
I see you've called this out under "Future Possibilities" below.

Reply:
Is this where we'd want to open on the `Installed` tab and and have the package in question selected potentially?

Reply:
That's the kind of thing I was thinking, yeah. It'd be useful regardless of the status indicators. Today, right clicking on a package node doesn't provide any way to get to the package manager. If you click on "Packages" you get "Manage NuGet packages". It'd be great to have a command on individual package nodes that opens the package manager with the package pre-selected.